code/game/dna/dna2.dm

```diff
@@ -102,112 +138,149 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+	SetDNAValueRange(DNA_UI_BODY_MARK_STYLE,	body_marks,		GLOB.marking_styles_list.len,		DNA_UI,	TRUE)
+	SetDNAValueRange(DNA_UI_TAIL_MARK_STYLE,	tail_marks,		GLOB.marking_styles_list.len,		DNA_UI,	TRUE)
 
 	//Set the Gender
 	switch(character.gender)
 		if(FEMALE)
-			SetUITriState(DNA_UI_GENDER, DNA_GENDER_FEMALE, 1)
+			SetDNATriState(DNA_UI_GENDER, DNA_GENDER_FEMALE, DNA_UI, TRUE)
 		if(MALE)
-			SetUITriState(DNA_UI_GENDER, DNA_GENDER_MALE, 1)
+			SetDNATriState(DNA_UI_GENDER, DNA_GENDER_MALE, DNA_UI, TRUE)
 		if(PLURAL)
-			SetUITriState(DNA_UI_GENDER, DNA_GENDER_PLURAL, 1)
+			SetDNATriState(DNA_UI_GENDER, DNA_GENDER_PLURAL, DNA_UI, TRUE)
 
 
-	UpdateUI()
+	UpdateDNA(DNA_UI)
 
-// Set a DNA UI block's raw value.
-/datum/dna/proc/SetUIValue(block, value, defer = FALSE)
-	if(block <= 0)
+
+/datum/dna/proc/ValidCheck(block, dna_type)
+	if(block <= 0 || dna_type < 0 || dna_type > 2 || dna_type == null)
+		return FALSE
+	else
+		return TRUE
+
+///////////////////////////////////////
+// SET and GET values for DNA blocks
+///////////////////////////////////////
+
+// Set a DNA block's raw value.
+/datum/dna/proc/SetDNAValue(block, value, dna_type, defer = FALSE)
+	if(!ValidCheck(block, dna_type))
 		return
 	ASSERT(value > 0)
 	ASSERT(value <= 4095)
-	UI[block]=value
-	dirtyUI = 1
+	switch(dna_type)
+		if(DNA_UI)
+			UI[block] = value
+			dirtyUI = TRUE
+		if(DNA_SE)
+			SE[block] = value
+			dirtySE = TRUE
+		if(DNA_RP)
+			RP[block] = value
+			dirtyRP = TRUE
 	if(!defer)
-		UpdateUI()
+		UpdateDNA(dna_type)
 
-// Get a DNA UI block's raw value.
-/datum/dna/proc/GetUIValue(block)
-	if(block <= 0)
+// Get a DNA block's raw value.
+/datum/dna/proc/GetDNAValue(block, dna_type)
+	if(!ValidCheck(block, dna_type))
 		return FALSE
-	return UI[block]
-
-// Set a DNA UI block's value, given a value and a max possible value.
-// Used in hair and facial styles (value being the index and maxvalue being the len of the hairstyle list)
-/datum/dna/proc/SetUIValueRange(block, value, maxvalue, defer = FALSE)
-	if(block <= 0)
+	switch(dna_type)
+		if(DNA_UI)
+			return UI[block]
+		if(DNA_SE)
+			return SE[block]
+		if(DNA_RP)
+			return RP[block]
+
+// Set a DNA block's value, given a value and a max possible value
+/datum/dna/proc/SetDNAValueRange(block, value, maxvalue, dna_type, defer = FALSE)
+	if(!ValidCheck(block, dna_type) || !value)
 		return
-	if(value == 0)
-		value = 1
 	ASSERT(maxvalue <= 4095)
 	var/range = (4095 / maxvalue)
-	if(value)
-		SetUIValue(block,round(value * range), defer)
+	switch(dna_type)
+		if(DNA_UI)			// Used in hair and facial styles (value being the index and maxvalue being the len of the hairstyle list)
+			if(value == 0)
+				value = 1
+			SetDNAValue(block, round(value * range), DNA_UI, defer)
+		if(DNA_SE)		// Might be used for species?
+			SetDNAValue(block, round(value * range) - rand(1, range - 1), DNA_SE)
+		if(DNA_RP)
+			SetDNAValue(block, round(value * range), DNA_RP)
```

Comment:
Great question. I'll need to brush up on the code before answering your question. I had the answer previously, but I need to find it again. :)

Reply:
This is a two-part question. 

- Q1: Is `defer` used for DNA_SE or DNA_RP?
- A1: No, `defer` is not used for DNA_SE nor DNA_RP.

```
if(!defer)
	UpdateDNA(DNA_UI)
```

- Q2: Even if it is not used, does quality coding dictate that we pass the argument along?
- A2: Ultimately, the `defer` argument is called in `ResetDNA`. This was previously two procs (ResetSE and ResetUI). In ResetUI, `defer` was called upon, in ResetSE, it was not. In order to preserve the argument, we pass it down through each proc that `ResetDNA` relies upon. This is why it is passed in `SetDNAValueRange` for DNA_UI.
We *can* pass it for DNA_SE and DNA_RP as well for a matter of consistency.

Reply:
The reason the code does not UpdateDNA when `!defer` is because if you are changing, for example, skin color for some species, you are changing _three_ DNA_UI values at once:

```
	SetDNAValueRange(DNA_UI_SKIN_R,		color2R(character.skin_colour),			255,	DNA_UI,	TRUE)
	SetDNAValueRange(DNA_UI_SKIN_G,		color2G(character.skin_colour),			255,	DNA_UI,	TRUE)
	SetDNAValueRange(DNA_UI_SKIN_B,		color2B(character.skin_colour),			255,	DNA_UI,	TRUE)
```
By ***deferring*** until all the three values are changed, the skin color for the mob sprite is updated appropriately instead of R, G, and B individually.

Reply:
@Kiyahitayika Hey smarty-pants, I think you're wrong.